### PR TITLE
DVC-6956: Refactor config structure

### PR DESCRIPTION
**Pull request by rob-odwyer**
This refactor should reduce the risk of bugs due to type casting in the segmentation code, plus (I think) make it a lot easier to follow.

tl;dr instead of having procedural `evaluateOperator` and `doesUserPassFilter` functions that manually check the types of the filters/operators and then navigate through the tree, every filter and operator type implements `FilterOrOperator`, meaning you can call `Evaluate` on it to check whether the whole tree under that point matches a given user and custom data.

### CODE REVIEW

File: bucketing/datamanager_platform.go [hunk 1]
```diff
@@ -6,7 +6,8 @@ func GetClientCustomData(sdkKey string) map[string]interface{} {
 	if data, ok := clientCustomData[sdkKey]; ok {
 		return data
 	}
-	return map[string]interface{}{}
+	// Nil maps are safe to read but not write, and this avoids an allocation
+	return nil
```
Comment (rob-odwyer) [inline bucketing/datamanager_platform.go:10]: I'm not really sure why this didn't get flagged as an allocation previously, or why these changes made it suddenly show up in the benchmarks.
This seems safe to do though, since we never write to the custom data map. The other option is returning a global empty map var, which seems sketchier if we ever did accidentally mutate it.

**Review by chris-hoefgen — APPROVED** [review]

File: bucketing/datamanager_platform.go (hunk 1 above)
Comment (chris-hoefgen) [reply bucketing/datamanager_platform.go:10]: at least returning nil we will know if it causes problems pretty quickly

File: bucketing/segmentation_test.go [hunk 2]
```diff
@@ -21,31 +21,27 @@ var brooks = api.PopulatedUser{
 	},
 }
 
+func _evaluateOperator(operator FilterOrOperator, audiences map[string]NoIdAudience, user api.PopulatedUser, clientCustomData map[string]interface{}) bool {
```
[the remark this answers is not in the document]
Comment (rob-odwyer) [reply bucketing/segmentation_test.go:24]: Yeah, TBH I was way too lazy to change all the `_evaluateOperator(...)` to `(...).evaluateOperator` in this file.
